content/library/advanced-features/app-testing/beyond-the-basics.md

```diff
@@ -0,0 +1,136 @@
+---
+title: Beyond the basics of app testing
+slug: /library/advanced-features/app-testing/beyond-the-basics
+---
+
+# Beyond the basics of app testing
+
+Now that you're comfortable with executing a basic test for a Streamlit app, let's cover the mutable attributes of [`AppTest`](/library/api-reference/app-testing/st.testing.v1.apptest):
+
+- `AppTest.secrets`
+- `AppTest.session_state`
+- `AppTest.query_params`
+
+For all three attributes, you can read and update values using dict-like syntax. You can use key notation but not attribute notation. For example, the `.secrets` attribute for `AppTest` accepts `at.secrets["my_key"]` but **_not_** `at.secrets.my_key`. This is differnt from how you can use the associated commands in the main library.
+
+For these attributes, the typical pattern is to declare any values before executing the first run of the app. Values can be inspected at any time in a test. There are a few extra considerations for secrets and Session State which we'll cover now.
+
+## Using secrets with app testing
+
+Be careful not to include secrets directly in your tests. Consider this simple project with `pytest` executed in the project's root directory:
+
+```none
+myproject/
+├── .streamlit/
+│   ├── config.toml
+│   └── secrets.toml
+├── app.py
+└── tests/
+    └── test_app.py
+```
+
+```bash
+cd myproject
+pytest tests/
+```
+
+In the above scenario, your simulated app will have access to your `secrets.toml` file. However, since you don't want to commit your secrets to your repository, you may need to write tests where you securely pull your secrets into memory or use dummy secrets.
+
+### Example: declaring secrets in a test
+
+Within a test, declare each secret after initializing your `AppTest` instance but before the first run. (A missing secret may result in an app that doesn't run!) For example, consider the following secrets file and corresponding test initialization to manually assign the same secrets:
+
+Secrets file:
+
+```toml
+db_username = "Jane"
+db_password = "mypassword"
+
+[my_other_secrets]
+things_i_like = ["Streamlit", "Python"]
+```
+
+Testing file with equivalent secrets:
+
+```python
+# Initialize an AppTest instance.
+at = AppTest.from_file("app.py")
+# Declare the secrets.
+at.secrets["db_username"] = "Jane"
+at.secrets["db_password"] = "mypassword"
+at.secrets["my_other_secrets.things_i_like"] = ["Streamlit", "Python"]
+# Run the app.
+at.run()
+```
+
+Generally, you don't want to type your secrets directly into your test. If you don't need your real secrets for a test, you can declare dummy secrets. If your tests require your real secrets, you should use an API to pass them securely and anonymously. If you are automating your tests with GitHub actions, check out their [Security guide](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions).
```

Comment:
Suggest talking about mocking explicitly here:

Generally, you don't want to type your secrets directly into your test. If you don't need your real secrets for a test, you can declare dummy secrets as in the example above. If your app uses secrets to connect to an external service like a database or API, consider mocking that service in your app tests. If you do need to use the real secrets and actually connect, you should use an API to pass them securely and anonymously. If you are automating your tests with GitHub actions, check out their [Security guide](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions).

Reply:
✅